### PR TITLE
Refactor compiler

### PR DESCRIPTION
**Pull request by aykevl**
Work-in-progress to refactor the compiler.

Not nearly finished, but sharing here for initial review to make sure this is the right direction. The main goal is to remove `Compiler.builder`, which is almost like a global variable (shared throughout the entire compiler instance). Instead, an IR builder should be made per function. This will make the structure of the compiler a lot simpler and should avoid lots of jumping around, for example to create wrapper functions. The last time I tried to remove the `SimpleDCE` mechanism (a blocker for building packages independently), wrapper functions were a major pain and resulted in needing to add a worklist for functions to be built after the current one is finished.

This change will be huge, so I decided to split it in many steps to make it reviewable. It could even be done in multiple PRs (if needed I could also split this PR up in multiple pieces). Most commits should be easy to review as the vast majority of the changes are mechanical (search+replace).

Some code is duplicated while the refactor is in progress. This duplication will be removed once it is finished.

### CODE REVIEW

**Comment by aykevl** [comment]
This requires #773 to be merged first, to make this a cleaner PR.

**Comment by aykevl** [comment]
Finally, this PR is ready for review. It doesn't fix the entire public interface of the compiler package (it still exposes `compiler.Phi` for example which is an internal detail) but it's a lot better now. But far more importantly, it removes the semi-global `llvm.Builder` used throughout the compiler and instead moves to a `compiler.builder`. This means that the state used during the compilation of a single function is now actually bound to a single function and not included in the semi-global `compilerContext` object.

It is a very large PR but it is split into many relatively small commits. The last time I tested, all but one commit can be compiled and tested individually, making it easy to spot regressions if there are any.

File: compiler/llvm.go [hunk 1]
```diff
@@ -29,34 +29,34 @@ func getUses(value llvm.Value) []llvm.Value {
...
+func (b *builder) emitPointerPack(values []llvm.Value) llvm.Value {
+	return llvmutil.EmitPointerPack(b.Builder, b.mod, b.Config, values)
 }
 
 // emitPointerUnpack extracts a list of values packed using emitPointerPack.
-func (c *Compiler) emitPointerUnpack(ptr llvm.Value, valueTypes []llvm.Type) []llvm.Value {
-	return llvmutil.EmitPointerUnpack(c.builder, c.mod, ptr, valueTypes)
+func (b *builder) emitPointerUnpack(ptr llvm.Value, valueTypes []llvm.Type) []llvm.Value {
+	return llvmutil.EmitPointerUnpack(b.Builder, b.mod, ptr, valueTypes)
 }
```
Comment (jaddr2line) [inline compiler/llvm.go]: You seem to be changing the terminology from `emit` to `create`. Should these be changed to make it consistent?

File: compiler/compiler.go [hunk 2]
```diff
@@ -364,26 +386,34 @@ func (c *Compiler) Compile(mainPath string) []error {
...
 // it as a LLVM type, creating it if necessary. It is a shorthand for
 // getLLVMType(getRuntimeType(name)).
-func (c *Compiler) getLLVMRuntimeType(name string) llvm.Type {
-	return c.getLLVMType(c.getRuntimeType(name))
+func (c *compilerContext) getLLVMRuntimeType(name string) llvm.Type {
+	fullName := "runtime." + name
+	typ := c.mod.GetTypeByName(fullName)
+	if typ.IsNil() {
+		println(c.mod.String())
+		panic("could not find runtime type: " + fullName)
```
Comment (jaddr2line) [inline compiler/compiler.go]: Again, try to use `fmt.Errorf`.

File: compiler/calls.go [hunk 3]
```diff
@@ -13,40 +11,35 @@ import (
...
-	runtimePkg := c.ir.Program.ImportedPackage("runtime")
-	member := runtimePkg.Members[fnName]
-	if member == nil {
-		panic("trying to call runtime." + fnName)
+// createCall creates a new call to runtime.<fnName> with the given arguments.
+func (b *builder) createRuntimeCall(fnName string, args []llvm.Value, name string) llvm.Value {
+	fullName := "runtime." + fnName
+	fn := b.mod.NamedFunction(fullName)
+	if fn.IsNil() {
+		panic("trying to call non-existing function: " + fullName)
```
[the remark this answers is not in the document]
Comment (aykevl) [reply compiler/calls.go]: Fixed the grammar.

I definitely agree that a real error message is better. However, that would also change lots of code so I want to do such refactoring at a later time to not complicate this one further.

File: compiler/llvm.go (hunk 1 above)
Comment (aykevl) [reply compiler/llvm.go]: Yes, but I think that's best done in a separate PR. And in this particular case I want `llvmutil.EmitPointerUnpack` and `(b *builder) emitPointerUnpack` to match.

File: compiler/compiler.go (hunk 2 above)
Comment (aykevl) [reply compiler/compiler.go]: Changing that here would imply changing lots of other places, which I think should be done separately. In fact, the panic here is already an improvement over the previous situation, which would probably panic with a nil pointer dereference or something similar.

**Comment by deadprogram** [comment]
I spent a while reviewing the changes, which indeed were quite numerous. That said, it was a true refactoring and not a rewrite, so although tedious to review, it was consistent. Thanks for that @aykevl 

I then turned my attentions to some empirical testing using my latest collection of demos plus a couple of test programs/boards I have been using lately. That also took a while, but all of the code/hardware functioned exactly as expected.

Therefore, I am going to merge this PR. Hopefully it will help current and future contributors navigate the compiler code.

**Pull request merged**